db/db_impl/db_impl_open.cc

```diff
@@ -1158,266 +1171,470 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& wal_numbers,
-          std::move(file), fname, immutable_db_options_.log_readahead_size,
-          io_tracer_, /*listeners=*/{}, /*rate_limiter=*/nullptr, is_retry));
-    }
-
-    // Create the log reader.
-    LogReporter reporter;
-    reporter.env = env_;
-    reporter.info_log = immutable_db_options_.info_log.get();
-    reporter.fname = fname.c_str();
-    reporter.old_log_record = &old_log_record;
-    if (!immutable_db_options_.paranoid_checks ||
-        immutable_db_options_.wal_recovery_mode ==
-            WALRecoveryMode::kSkipAnyCorruptedRecords) {
-      reporter.status = nullptr;
-    } else {
-      reporter.status = &status;
-    }
-    // We intentially make log::Reader do checksumming even if
-    // paranoid_checks==false so that corruptions cause entire commits
-    // to be skipped instead of propagating bad information (like overly
-    // large sequence numbers).
-    log::Reader reader(immutable_db_options_.info_log, std::move(file_reader),
-                       &reporter, true /*checksum*/, wal_number);
-
-    // Determine if we should tolerate incomplete records at the tail end of the
-    // Read all the records and add to a memtable
-    std::string scratch;
-    Slice record;
-
-    const UnorderedMap<uint32_t, size_t>& running_ts_sz =
-        versions_->GetRunningColumnFamiliesTimestampSize();
-
-    TEST_SYNC_POINT_CALLBACK("DBImpl::RecoverLogFiles:BeforeReadWal",
-                             /*arg=*/nullptr);
-    uint64_t record_checksum;
-    while (!stop_replay_by_wal_filter &&
-           reader.ReadRecord(&record, &scratch,
-                             immutable_db_options_.wal_recovery_mode,
-                             &record_checksum) &&
-           status.ok()) {
-      if (record.size() < WriteBatchInternal::kHeader) {
-        reporter.Corruption(record.size(),
-                            Status::Corruption("log record too small"));
-        continue;
-      }
-      // We create a new batch and initialize with a valid prot_info_ to store
-      // the data checksums
-      WriteBatch batch;
-      std::unique_ptr<WriteBatch> new_batch;
+    bool read_record = reader->ReadRecord(
+        &record, &scratch, immutable_db_options_.wal_recovery_mode,
+        &record_checksum);
 
-      status = WriteBatchInternal::SetContents(&batch, record);
-      if (!status.ok()) {
-        return status;
-      }
+    // FIXME(hx235): consolidate `read_record` and `status`
+    if (!read_record || !status.ok()) {
+      break;
+    }
 
-      const UnorderedMap<uint32_t, size_t>& record_ts_sz =
-          reader.GetRecordedTimestampSize();
-      status = HandleWriteBatchTimestampSizeDifference(
-          &batch, running_ts_sz, record_ts_sz,
-          TimestampSizeConsistencyMode::kReconcileInconsistency, seq_per_batch_,
-          batch_per_txn_, &new_batch);
-      if (!status.ok()) {
-        return status;
-      }
+    // FIXME(hx235): consolidate `process_status` and `status`
```

Comment:
I have some ideas for this, we can sync offline.

Reply:
Tracked